src/field_theory/splitting_field.lean

```diff
@@ -208,6 +229,74 @@ end
 
 end splits
 
+end polynomial
+
+
+section embeddings
+
+variables (F : Type*) [field F]
+
+/-- The bottom subalgebra is isomorphic to the field. -/
+def alg_equiv.of_bot (R : Type*) [semiring R] [nontrivial R] [algebra F R] :
+  (⊥ : subalgebra F R) ≃ₐ[F] F :=
+alg_equiv.symm $ alg_equiv.of_bijective (algebra.of_id F _)
+⟨ring_hom.injective _, λ ⟨y, hy⟩, let ⟨x, hx⟩ := algebra.mem_bot.1 hy in ⟨x, subtype.eq hx⟩⟩
+
+/-- If `p` is the minimal polynomial of `a` over `F` then `F[a] ≃ₐ[F] F[x]/(p)` -/
+def alg_equiv.adjoin_singleton {R : Type*} [comm_ring R] [algebra F R]
+  (x : R) (hx : is_integral F x) :
+  algebra.adjoin F ({x} : set R) ≃ₐ[F] adjoin_root (minimal_polynomial hx) :=
+alg_equiv.symm $ alg_equiv.of_bijective
+  (alg_hom.cod_restrict
+    (adjoin_root.alg_hom _ x $ minimal_polynomial.aeval hx) _
+    (λ p, adjoin_root.induction_on _ p $ λ p,
+      (algebra.adjoin_singleton_eq_range F x).symm ▸ ⟨p, rfl⟩))
+  ⟨(alg_hom.injective_cod_restrict _ _ _).2 $ (alg_hom.injective_iff _).2 $ λ p,
+    adjoin_root.induction_on _ p $ λ p hp, ideal.quotient.eq_zero_iff_mem.2 $
+    ideal.mem_span_singleton.2 $ minimal_polynomial.dvd hx hp,
+  λ y, let ⟨p, hp⟩ := (subalgebra.ext_iff.1 (algebra.adjoin_singleton_eq_range F x) y).1 y.2 in
+  ⟨adjoin_root.mk _ p, subtype.eq hp⟩⟩
+
+open finset
+
+-- Why is this so slow?
+/-- If the minimal polynomial of each `ai` splits in `L` then `F(a1, ..., an)` embeds in `L`. -/
```

Comment:
I don't understand this docstring. What are `ai` and `F`?

Reply:
done